### PR TITLE
engine: improve error handling and retry logic in portforwarding. Fixes #2099

### LINKED ISSUE #2099: [ BUGish ] Port-forward breaks, doesn't come back
Been playing with Tilt today and I realized one thing. The port forward sometimes disappear and when this happens, there's no way to get it back other than `CTRL-C` and `tilt up`, is there anyway to have a keepalive running ?

Running tilt locally to our Kubernetes cluster in AWS.

### PR DESCRIPTION
**Pull request by nicks**
Hello @landism, @maiamcc,

Please review the following commits I made in branch nicks/issue2099:

1b708ecd8b67a8c953bf3f26a108d36b0e9f4668 (2019-09-13 15:48:00 -0400)
engine: improve error handling and retry logic in portforwarding. Fixes #2099

### CODE REVIEW

File: internal/k8s/portforward.go [hunk 1]
```diff
@@ -19,30 +19,47 @@ import (
 )
 
 type PortForwardClient interface {
-	Create(ctx context.Context, namespace Namespace, podID PodID, localPort int, remotePort int) (closer func(), err error)
+	// Creates a new port-forwarder that's bound to the given context's lifecycle.
+	// Then the context is canceled, the port-forwarder will close.
```
Comment (maiamcc) [inline internal/k8s/portforward.go]:
```suggestion
	// When the context is canceled, the port-forwarder will close.
```

File: internal/k8s/portforward.go [hunk 2]
```diff
@@ -19,30 +19,47 @@ import (
...
+	Create(ctx context.Context, namespace Namespace, podID PodID, localPort int, remotePort int) (PortForwarder, error)
```
Comment (maiamcc) [inline internal/k8s/portforward.go]: Better named `NewPortForwarderer` or something, no?

File: internal/engine/portforwardcontroller.go [hunk 3]
```diff
@@ -100,26 +102,73 @@ func (m *PortForwardController) OnChange(ctx context.Context, st store.RStore) {
...
+		err := m.onePortForward(ctx, entry, forward)
+		if ctx.Err() != nil {
+			// If the context was canceled, we're satisfied.
+			// Ignore any errors.
+			return
+		}
+
+		// Otherwise, repeat the loop, maybe logging the error
+		if err != nil {
+			logger.Get(ctx).Infof("Error port-forwarding %s: %v", entry.name, err)
```
Comment (landism) [inline internal/engine/portforwardcontroller.go]: Maybe add a "Reconnecting..." here, to reduce the chance a user thinks they need to take action

File: internal/engine/portforwardcontroller.go [hunk 4]
```diff
@@ -100,26 +102,73 @@ func (m *PortForwardController) OnChange(ctx context.Context, st store.RStore) {
...
+			currentBackoff = originalBackoff
 		}
 	}
 }
 
+func (m *PortForwardController) onePortForward(ctx context.Context, entry portForwardEntry, forward model.PortForward) error {
+	ns := entry.namespace
+	podID := entry.podID
+
+	// TODO(nick): Add exponential retry & backoff
```
Comment (landism) [inline internal/engine/portforwardcontroller.go]: is this still relevant, or is it taken care of by startPortForwardLoop?

File: internal/k8s/portforward.go (hunk 1 above)
Comment (nicks) [reply internal/k8s/portforward.go]: done

File: internal/k8s/portforward.go (hunk 2 above)
Comment (nicks) [reply internal/k8s/portforward.go]: done

File: internal/engine/portforwardcontroller.go (hunk 3 above)
Comment (nicks) [reply internal/engine/portforwardcontroller.go]: done

File: internal/engine/portforwardcontroller.go (hunk 4 above)
Comment (nicks) [reply internal/engine/portforwardcontroller.go]: done

**Pull request merged**